Merge utils/Makefile into the root Makefile

shindere opened the pull request:
The main thing this PR does is to modify the configure and build systems
so that `ocamltest/ocamltest_config.ml` and `utils/config.ml` (renamed to
`utils/config_fragment.ml` by this PR) are generated as part of the
configure stager, rather than during the build.

This PR consists in several commits which are best reviewed one after the
other.



Two questions regarding the configure module:

 1. Wouldn't it be a good moment to remove `bytecode_c_compiler` and the like, or do those need to be deprecated first?

 2. The code that computes `mkexe`, `mkdll` and `mkmaindll` is executed at runtime. Is that a feature, i.e. a behaviour that needs to be preserved as it it, or could we save a test and a few computations and generate the correct code at configure time?

shindere committed:
Provide an expanded value for BINDIR and LIBDIR during configuration

This will be used to generate the OCaml configuration module as part of
the configure stage, rather than generating it during the build.

shindere commented:
Is it okay to stop generating `utils/config_boot.mli` and
`utils/config_main.mli` as the PR does?

These files look not so useful since the corresponding implementations
won't actually be compiled as such but rather copied to `config.ml`
which will then be compiled and checked against `config.mli`.

dra27 commented:
> Is it okay to stop generating `utils/config_boot.mli` and `utils/config_main.mli` as the PR does?
> 
> These files look not so useful since the corresponding implementations won't actually be compiled as such but rather copied to `config.ml` which will then be compiled and checked against `config.mli`.

These files are there so that if you add, say, `Config.my_amazing_new_feature_flag` to `utils/config.mli` and `utils/config.mlp`, then during the build you get:

```
File "/home/dra/ocaml/utils/config_boot.ml", line 1:
Error: The implementation utils/config_boot.ml
       does not match the interface utils/config_boot.cmi:
       The value `my_amazing_new_feature_flag' is required but not provided
       File "/home/dra/ocaml/utils/config_boot.mli", line 262, characters 0-17:
         Expected declaration
```

Otherwise you only find this out when `make bootstrap` is next run (hopefully in CI, of course, but nonetheless further down the line).

gasche commented:
> These files are there so that [..]

This reads like a design aspect that could be documented by a comment somewhere in the configuration / build system. (Maybe the PR could include such a comment?)

dra27 commented:
> > These files are there so that [..]
> 
> This reads like a design aspect that could be documented by a comment somewhere in the configuration / build system. (Maybe the PR could include such a comment?)

Guilty as charged! It could/should indeed be documented with the definition of `ALL_CONFIG_CMO` in `compilerlibs/Makefile.compilerlibs`.

shindere commented:
David Allsopp (2022/05/24 02:59 -0700):
> > Is it okay to stop generating `utils/config_boot.mli` and `utils/config_main.mli` as the PR does?
> >
> > These files look not so useful since the corresponding implementations won't actually be compiled as such but rather copied to `config.ml` which will then be compiled and checked against `config.mli`.
>
> These files are there so that if you add, say, `Config.my_amazing_new_feature_flag` to `utils/config.mli` and `utils/config.mlp`, then during the build you get:
>
> ```
> File "/home/dra/ocaml/utils/config_boot.ml", line 1:
> Error: The implementation utils/config_boot.ml
>        does not match the interface utils/config_boot.cmi:
>        The value `my_amazing_new_feature_flag' is required but not provided
>        File "/home/dra/ocaml/utils/config_boot.mli", line 262, characters 0-17:
>          Expected declaration
> ```
>
> Otherwise you only find this out when `make bootstrap` is next run
> (hopefully in CI, of course, but nonetheless further down the line).

I'll try to do it with the recently added `-cmi-file` compiler option
which should avoid the need to cpy files around.

shindere commented:
Gabriel Scherer (2022/05/24 03:09 -0700):
> > These files are there so that [..]
>
> This reads like a design aspect that could be documented by a comment
> somewhere in the configuration / build system. (Maybe the PR could
> include such a comment?)

Let's first see whether we can get rid of it or not and then, yes,
definitely, if it stays in place it will be worth documenting it. but I
hope we can get rid of it in an elegant manner.

shindere commented:
Although there was no conflict with trunk it felt a good idea to rebase.

shindere commented:
Just rebased on latest trunk to resolve merge conflicts

dra27 commented on Makefile.config.in:
```diff
@@ -97,7 +97,7 @@ O=@OBJEXT@
 
 ### How to tell the C compiler to output an object file
-OUTPUTOBJ=@outputobj@
+OUTPUTOBJ=@outputobj@ $(EMPTY)
```
The space before `$(EMPTY)` here needs to go.

dra27 replied:
The space before `$(EMPTY)` shouldn't be here (same in `OUTPUTEXE` on L192)

dra27 replied:
The space before `$(EMPTY)` shouldn't be here (same in `OUTPUTEXE` on L192)

dra27 commented on configure.ac:
```diff
@@ -1220,13 +1281,16 @@ AS_IF([test -z "$PARTIALLD"],
-    [PACKLD="link -lib -nologo $machine -out:"],
-    [PACKLD="$DIRECT_LD -r$PACKLD_FLAGS -o \$(EMPTY)"])],
-  [PACKLD="$PARTIALLD -o \$(EMPTY)"])
+    [pacld_ocaml="link -lib -nologo $machine -out:"
```
Typo: this should be `packld_ocaml` (missing k) - although suggestion earlier that this separate variable isn't needed.

dra27 commented on configure.ac:
```diff
@@ -169,10 +200,14 @@ AC_SUBST([function_sections])
 AC_SUBST([flexdll_chain])
+AC_SUBST([mkdll_ldflags_exp])
 AC_SUBST([PACKLD])
+AC_SUBST([packld_ocaml])
```
I don't think this variable is needed - `$(EMPTY)` is used because some linkers _must_ have a space after `-o`. There's no harm for the Microsoft case having configure set `PACKLD="link -lib -nologo $machine -out:"` (no space at the end) and then have in `Makefile.config.in` `PACKLD=@PACKLD@$(EMPTY)`? `@PACKLD@` can then be used directly in `utils/config_fragment.ml.in`

dra27 commented on configure.ac:
```diff
@@ -591,19 +652,18 @@ AS_IF(
 
 ## Determine which flags to use for the C compiler
 
+outputobj='-o '
```
I find this a little inconsistent: in the `AS_CASE` below, we very clearly set all `outputobj`, `warn_error_flag`, and `cc_warnings` for various different configurations.

If the default for `outputobj` goes here shouldn't the defaults for `warn_error_flag` and `cc_warnings` go here too?

dra27 commented on configure.ac:
```diff
@@ -525,29 +565,50 @@ AS_CASE([$ocaml_cv_cc_vendor],
+
+# Absolute path to the toplevel source directory
+
+# A sentinelle character (X) is added ad the end of the directory name
```
Two typos: _at_ the end of the directory name and _sentinel_, rather than sentinelle. I would be more explicit on the _exact_ reason for the sentinel: it's there because `$(..)` strips trailing `\n` characters (you could reference Eric's marvellous reply in https://lists.gnu.org/archive/html/autoconf/2019-07/msg00002.html !)

dra27 commented on configure.ac:
```diff
@@ -525,29 +565,50 @@ AS_CASE([$ocaml_cv_cc_vendor],
 AS_CASE([$host],
   [*-*-mingw32|*-pc-windows],
     [unix_or_win32="win32"
+    ocamlsrcdir="$(LC_ALL=C.UTF-8 cygpath -w "$ocamlsrcdir")"
```
Needs to be a `--` after the `-w`

dra27 commented on configure.ac:
```diff
@@ -525,29 +565,50 @@ AS_CASE([$ocaml_cv_cc_vendor],
+# A sentinelle character (X) is added ad the end of the directory name
+# and then removed, in case it finishes with an odd character.
+
+ocamlsrcdir=$(unset CDPATH; cd -- "$srcdir" && printf %sX "$PWD") || fail
```
I would mention here in a comment that we use `printf` rather than `echo` because `echo` has indeterminate and uncontrollable behaviour if `$PWD` contains backslashes.

dra27 commented on configure.ac:
```diff
@@ -223,14 +260,17 @@ AS_CASE([$host],
     ccomptype=msvc
+    CSC=csc
+    CSCFLAGS="/nologo /nowarn:1668"
+    AS_CASE([$host_cpu], [i*86], [CSCFLAGS="$CSCFLAGS /platform:x86"])
```
This wants to be in a separate `AS_CASE` of its own, as it should be set for _both_ mingw and msvc

dra27 commented on configure.ac:
```diff
@@ -525,29 +565,50 @@ AS_CASE([$ocaml_cv_cc_vendor],
   [sunc-*],
     [CPP="$CC -E -Qn"], # suppress generation of Sun PRO ident string
```
`ocamltest_CPP` needs to be set to `"$CPP"` here and in the previous case

dra27 commented on configure.ac:
```diff
@@ -1831,11 +1895,14 @@ AC_CHECK_FUNC([_BitScanForward], [AC_DEFINE([HAS_BITSCANFORWARD])])
 AS_CASE([$enable_debugger],
   [no],
     [with_debugger=""
+    ocamltest_ocamldebug=false
```
I was going to suggest that it's a shame that we end up with `with_debugger` and `ocamltest_ocamldebug` (and similarly for `ocamltest_ocamldoc`), which it is... however, as they end up in `Makefile.config`, we probably need to keep both, even if we stop the root `Makefile` from using the `$with_debugger` and `$ocamldoc`, so fiddling with those can wait for a future PR! :)

dra27 commented on ocamltest/Makefile:
```diff
@@ -282,7 +167,7 @@ ocamltest.html: ocamltest.org
 clean:
 	rm -rf ocamltest ocamltest.exe ocamltest.opt ocamltest.opt.exe
-	rm -rf $(c_files:.c=.o) $(c_files:.c=.obj)
+	rm -rf *.o *.obj
```
I can see why this has changed, but therefore the line below is no longer needed. If `.o` files are deleted using a wildcard, might we as well delete the OCaml artefacts using a wildcard, too?

dra27 commented on Makefile.build_config.in:
```diff
@@ -55,15 +55,17 @@ DOCUMENTATION_TOOL_CMD=@documentation_tool_cmd@
+ocamltest_unix_impl = @ocamltest_unix_impl@
+ocamltest_unix_name = @ocamltest_unix_name@
+ocamltest_unix_path = @ocamltest_unix_path@
+ocamltest_unix_include = @ocamltest_unix_include@
```
`$(EMPTY)` could be added to the end of this (with no space) to allow it to be removed from `configure.ac`

dra27 commented on configure.ac:
```diff
@@ -61,6 +66,16 @@ instrumented_runtime=true
+lib_str=false
+lib_systhreads=false
+lib_unix=false
+ocamltest_libunix=None
```
There's a bit of a problem with this, and I'm not sure which solution you'll want to do. At the moment, this variable should be `None` if the Unix library is not built (as here) but then `Some true` for Unix systems and `Some false` for Windows systems (the `Some false` value is missing). This is used to implement the `libunix` and `libwin32unix` actions in the testsuite. The simplest fix here is to implement the `Some false` case, but the reality is that it would be better to eliminate the `libunix` and `libwin32unix` actions, sticking with `hasunix` followed by `windows` or `not-windows` as required. What do you think?

dra27 commented on configure.ac:
```diff
@@ -61,6 +66,16 @@ instrumented_runtime=true
+lib_unix=false
+ocamltest_libunix=None
+ocamltest_unix_impl="dummy"
+ocamltest_unix_name=""
+ocamltest_unix_path=""
+ocamltest_unix_include=""
```
For my taste, these three separate variables go a little bit too far? Isn't it OK for `ocamltest/Makefile` to know that the Unix library's basename is `unix` and that if the path is given, we need to include it? In that case, `ocamltest_unix_path` could be set to `'/otherlibs/unix'` (note no `$(ROOTDIR)`) or `''` and `ocamltest/Makefile` could use that single value to instantiate `-I $(ROOTDIR)$(ocamltest_unix_path)`?

dra27 commented on configure.ac:
```diff
@@ -525,29 +565,50 @@ AS_CASE([$ocaml_cv_cc_vendor],
+   ocamltest_unix_impl="real"
+   ocamltest_unix_name="unix"
+   ocamltest_unix_path='$(ROOTDIR)/otherlibs/unix'
+   ocamltest_unix_include='-I $(ROOTDIR)/otherlibs/unix $(EMPTY)'])
```
If these are kept, the `$(EMPTY)` could go in `Makefile.build_config.in` (with no space before it) and the trailing space be left here, which removes more make macros from `configure.ac`. `$(ROOTDIR)` could be added to the others using `$(ocamltest_unix_path:%=$(ROOTDIR)%)` (which is empty if `$(ocamltest_unix_path)` is empty, but otherwise prefixes it with `$(ROOTDIR)`.

dra27 commented on configure.ac:
```diff
@@ -61,6 +66,16 @@ instrumented_runtime=true
 force_instrumented_runtime=false
 instrumented_runtime_libs=""
 bootstrapping_flexdll=false
+flexdll_dirs=[[]]
```
Is there a subtle reason for using m4sh escaping (`[[]]`) rather than sh escaping (`'[]'`)?

dra27 commented on utils/config_fragment.ml.in:
```diff
@@ -0,0 +1,106 @@
+
+let systhread_supported = @systhread_support@
+
+let flexdll_dirs = @flexdll_dirs@
```
This is always a list, so it might be clearer to initialise `flexdll_dirs` to `""` in `configure.ac` and then set it `'"+flexdll"'` when bootstrapping and here have `[@flexdll_dirs@]` - or do the square brackets do something strange with the expander??

dra27 commented on Makefile.build_config.in:
```diff
@@ -55,15 +55,17 @@ DOCUMENTATION_TOOL_CMD=@documentation_tool_cmd@
 # Git submodule)
 FLEXDLL_SOURCES=@flexdir@
 BOOTSTRAPPING_FLEXDLL=@bootstrapping_flexdll@
+FLEXDLL_DIR=@flexdll_dir@
```
This isn't needed (if it's kept, it needs the 's' added to `@flexdll_dirs@`)

dra27 commented on utils/config_fragment.ml.in:
```diff
@@ -0,0 +1,106 @@
+    let flexlink_flags = {@QS@|@mkexe_extra_flags@|@QS@} in
+    let flags = " -chain " ^ flexdll_chain ^ " " ^ flexlink_flags in
+    flexlink ^ flags ^ {@QS@|@mkdll_ldflags_exp@|@QS@},
+    flexlink ^ " -exe" ^ flags ^ {@QS@|@mkdll_ldflags_exp@|@QS@},
```
This one should be using OC_LDFLAGS, not `OC_DLL_LDFLAGS`. I think an equivalent `@mkexe_ldflags_exp@` is needed?

dra27 commented on configure.ac:
```diff
@@ -182,6 +217,8 @@ AC_CONFIG_FILES([Makefile.config])
 AC_CONFIG_FILES([manual/src/version.tex])
 AC_CONFIG_FILES([manual/src/html_processing/src/common.ml])
+AC_CONFIG_FILES([ocamltest/ocamltest_config.ml])
+AC_CONFIG_FILES([utils/config_fragment.ml])
```
Minor file-naming bike-shedding: we have `utils/config.fixed.ml` already - `utils/config.fragment.ml` or even `utils/config.generated.ml`?

dra27 replied:
Minor file-naming bike-shedding: we have `utils/config.fixed.ml` already - `utils/config.fragment.ml` or even `utils/config.generated.ml`?

dra27 commented on utils/Makefile:
```diff
@@ -1,127 +0,0 @@
-#**************************************************************************
```
The `SUBST` and `SUBST_STRING` macros in `Makefile.common` can also be deleted along with this file!

dra27 commented on ocamltest/ocamltest_config.ml.in:
```diff
@@ -15,77 +17,77 @@
+let libunix = @ocamltest_libunix@
 
-let systhreads = %%systhreads%%
+let systhreads = @lib_systhreads@
```
`@systhread_support@` could be used, and the extra variable `lib_systhreads` eliminated.

dra27 commented on configure.ac:
```diff
@@ -2138,18 +2221,20 @@ AC_CONFIG_COMMANDS_PRE([
         mkdll_ldflags="${mkdll_ldflags} ${mkexe_ldflags_prefix}${flag}"
-      done])
+      done
+      mkdll_ldflags_exp="$mkdll_ldflags"])
```
I think a `mkexe_ldflags_exp` is needed here, too (see later comment in `utils/config_fragment.ml.in`. It would want the oc_ldflags part below on L2231 as well but then _not_ the oc_dll_ldflags bit on L2237

dra27 reviewed (commented):
It is really great to see the back of those `sed` expressions, and sorry it's taken me a while to review this. I've checked the various files generated commit-by-commit to try to try to catch the effect of all the moving parts.

A few of the comments are suggestions for clarifications or tweaks, but they're mostly optional: in particular, I think possibly a little too much has been removed from `ocamltest/Makefile` into `configure.ac` and there's a slightly muddying of the proverbial waters because there's some extant win32unix/unix code in ocamltest which you might want to excise at the same time.

shindere committed:
ocamltest: determine ocamlsrcdir during configure rather than during build

shindere committed:
ocamltest: determine C#-related information during configure rather than build

shindere committed:
Generate ocamltest/ocamltest_config.ml during configure rather than build

shindere committed:
Introduce config.generated.ml

This is the fragment of the configure module which is presently generated
at build time from a template file but will ultimately be generated
at configure time.

Also make .gitignore more specific and explicit about what needs
to be ignored.

shindere committed:
Merge utils/Makefile into the root Makefile

shindere commented:
Hey David,

Many thanks for your review, precious as usually!

David Allsopp (2022/07/12 01:08 -0700):
>    @dra27 commented on this pull request.
> 
>    It is really great to see the back of those sed expressions,

:-)

>    and sorry it's taken me a while to review this.

No problem. It seems it has taken me a while to take your review into
account, too! Let's hope this can now be merged soon, to avoid the
rebase dances which are both time-consuming and error-prone.

>    I've checked the various files
>    generated commit-by-commit to try to try to catch the effect of all the
>    moving parts.

I did try to keep the history coherent so I really appreciate this
commit-per-commit review, many thanks for that!

Out of curiousity: how easy would it be to also perform the review
itself on a commit-per-commit basis to make it clearer to which commit
each comment is related? Would that make sense in GitHub's interface?
Does the interface make it possible (easy?) to do reviews in this way?

If not, it's not a big deal. I was asking mainly out of curiosity.

>    A few of the comments are suggestions for clarifications or tweaks, but
>    they're mostly optional: in particular, I think possibly a little too
>    much has been removed from ocamltest/Makefile into configure.ac and
>    there's a slightly muddying of the proverbial waters because there's
>    some extant win32unix/unix code in ocamltest which you might want to
>    excise at the same time.

I didn't dare to, so far.

>      __________________________________________________________________
> 
>    In [1]Makefile.config.in:
> > @@ -97,7 +97,7 @@ ***@***.***@
>  ***@***.***@
> 
>  ### How to tell the C compiler to output an object file
> ***@***.***@
> ***@***.***@ $(EMPTY)
> 
>    The space before $(EMPTY) here needs to go.

Yes of course you are obviously correct, many thanks. The space was
actually defeating the whole purpose of this change, nothing less. It's
fixed.

>      __________________________________________________________________
> 
>    In [2]configure.ac:
> >      # For the Microsoft C compiler there must be no space at the end of the
>      # string.
> -    [PACKLD="link -lib -nologo $machine -out:"],
> -    [PACKLD="$DIRECT_LD -r$PACKLD_FLAGS -o \$(EMPTY)"])],
> -  [PACKLD="$PARTIALLD -o \$(EMPTY)"])
> +    [pacld_ocaml="link -lib -nologo $machine -out:"
> 
>    Typo: this should be packld_ocaml (missing k) - although suggestion
>    earlier that this separate variable isn't needed.

Fixed.

>      __________________________________________________________________
> 
>    In [3]configure.ac:
> >  AC_SUBST([PACKLD])
> +AC_SUBST([packld_ocaml])
> 
>    I don't think this variable is needed - $(EMPTY) is used because some
>    linkers must have a space after -o. There's no harm for the Microsoft
>    case having configure set PACKLD="link -lib -nologo $machine -out:" (no
>    space at the end) and then have in Makefile.config.in
>    ***@***.***@$(EMPTY)? @PACKLD@ can then be used directly in
>    utils/config_fragment.ml.in

Change done. Thanks.

>      __________________________________________________________________
> 
>    In [4]configure.ac:
> > @@ -591,19 +652,18 @@ AS_IF(
> 
>  ## Determine which flags to use for the C compiler
> 
> +outputobj='-o '
> 
>    I find this a little inconsistent: in the AS_CASE below, we very
>    clearly set all outputobj, warn_error_flag, and cc_warnings for various
>    different configurations.
> 
>    If the default for outputobj goes here shouldn't the defaults for
>    warn_error_flag and cc_warnings go here too?

You are right that this change introduced some inconsistency which was
not there before, sorry about that. I removed that commit. What I had in
mind is that, ultimately, we could adopt a style where we make sure that
all the variables we use in `configure` are initialized. This change was
a step in this ddirection but it's probably better to change all the
variables at once.

In particular, currently we set-up our configuration variables in two
steps. We use the `AC_SUBST` macro to let `configure.status` substitute
them in various places and we (sometimes, but not always) initialize
them.

I think it would be better style to introduce an m4 macro,
`OCAML_CONFIG_VAR`, say, that would perform both tasks: calling
`AC_SUBST` and initialize the variable with a decent value. This would
make `configure.ac` slightly shorter and, more importantly, more
disciplined. We could even have a CI test to ensure that `configure.ac`
never calls `AC_SUBST` directly.

>    In [5]configure.ac:
> >  #  TODO: why can we not use $CPP in multicore, fix this?
> -  [CPP="$CC -E -P"])
> +  [CPP="$CC -E -P"
> +  ocamltest_CPP="$CPP"])
> +
> +# Absolute path to the toplevel source directory
> +
> +# A sentinelle character (X) is added ad the end of the directory name
> 
>    Two typos: at the end of the directory name and sentinel, rather than
>    sentinelle. I would be more explicit on the exact reason for the
>    sentinel: it's there because $(..) strips trailing \n characters (you
>    could reference Eric's marvellous reply in
>    [6]https://lists.gnu.org/archive/html/autoconf/2019-07/msg00002.html
>    !)

Done.

>      __________________________________________________________________
> 
>    In [7]configure.ac:
> >
>  # Libraries to build depending on the host
> 
>  AS_CASE([$host],
>    [*-*-mingw32|*-pc-windows],
>      [unix_or_win32="win32"
> +    ocamlsrcdir="$(LC_ALL=C.UTF-8 cygpath -w "$ocamlsrcdir")"
> 
>    Needs to be a -- after the -w

Fixed

>      __________________________________________________________________
> 
>    In [8]configure.ac:
> >  #  TODO: why can we not use $CPP in multicore, fix this?
> -  [CPP="$CC -E -P"])
> +  [CPP="$CC -E -P"
> +  ocamltest_CPP="$CPP"])
> +
> +# Absolute path to the toplevel source directory
> +
> +# A sentinelle character (X) is added ad the end of the directory name
> +# and then removed, in case it finishes with an odd character.
> +
> +ocamlsrcdir=$(unset CDPATH; cd -- "$srcdir" && printf %sX "$PWD") || fail
> 
>    I would mention here in a comment that we use printf rather than echo
>    because echo has indeterminate and uncontrollable behaviour if $PWD
>    contains backslashes.

I documented this.

>    In [9]configure.ac:
> > @@ -223,14 +260,17 @@ AS_CASE([$host],
>    [*-pc-windows],
>      [CC=cl
>      ccomptype=msvc
> +    CSC=csc
> +    CSCFLAGS="/nologo /nowarn:1668"
> +    AS_CASE([$host_cpu], [i*86], [CSCFLAGS="$CSCFLAGS /platform:x86"])
> 
>    This wants to be in a separate AS_CASE of its own, as it should be set
>    for both mingw and msvc

Fixed. Thanks.

>    In [10]configure.ac:
> > @@ -525,29 +565,50 @@ AS_CASE([$ocaml_cv_cc_vendor],
>    [sunc-*],
>      [CPP="$CC -E -Qn"], # suppress generation of Sun PRO ident string
> 
>    ocamltest_CPP needs to be set to "$CPP" here and in the previous
>    case

Well spotted, thanks! This has now been fixed.

>    In [11]configure.ac:
> > @@ -1831,11 +1895,14 @@ AC_CHECK_FUNC([_BitScanForward], [AC_DEFINE([HAS_BITSC
> ANFORWARD])])
>  AS_CASE([$enable_debugger],
>    [no],
>      [with_debugger=""
> +    ocamltest_ocamldebug=false
> 
>    I was going to suggest that it's a shame that we end up with
>    with_debugger and ocamltest_ocamldebug (and similarly for
>    ocamltest_ocamldoc), which it is... however, as they end up in
>    Makefile.config, we probably need to keep both, even if we stop the
>    root Makefile from using the $with_debugger and $ocamldoc, so fiddling
>    with those can wait for a future PR! :)

Yes, I am definitely looking forward to have these non-boolean variables
to disappear. At least, I think we should confine them to
`Makefile.config`, deprecate them and, ultimately, remove them.

>    In [12]ocamltest/Makefile:
> > @@ -282,7 +167,7 @@ ocamltest.html: ocamltest.org
>  .PHONY: clean
>  clean:
>         rm -rf ocamltest ocamltest.exe ocamltest.opt ocamltest.opt.exe
> -       rm -rf $(c_files:.c=.o) $(c_files:.c=.obj)
> +       rm -rf *.o *.obj
> 
>    I can see why this has changed, but therefore the line below is no
>    longer needed. If .o files are deleted using a wildcard, might we as
>    well delete the OCaml artefacts using a wildcard, too?

Yes, absolutely. Done. thanks.

Actually, when I first wrote this Makefile, I had the opinion that it
should control everything as precisely as possible, which menat to
exclude the use of patterns like `*.o`. Now I think one should
distinguish building and installing, on the one hand, from cleaning, on
the other hand. I still think tha in the first case it's not a good
practice to use generic patterns and that it's a cleaner approach to be
explicit about what you build and install, because, among other things,
this opens the way to providing an `uninstall` target. But for cleaning
and removing I agree that it's important to remove whatever artifact
there is and that, in this situation, using patterns is fine, or even
better than using variables would leave some files undeleted when a
renaming occurs.

>    In [13]Makefile.build_config.in:
> >
>  ### Where to install documentation
>  PACKAGE_TARNAME = @PACKAGE_TARNAME@
>  datarootdir = @datarootdir@
>  ***@***.***@
> 
> -# Expanded form of MKEXE_EXP and MKDLL_EXP _with host flexlink command_
> ***@***.***_exp@
> ***@***.***_exp@
> +ocamltest_unix_impl = @ocamltest_unix_impl@
> +ocamltest_unix_name = @ocamltest_unix_name@
> +ocamltest_unix_path = @ocamltest_unix_path@
> +ocamltest_unix_include = @ocamltest_unix_include@
> 
>    $(EMPTY) could be added to the end of this (with no space) to allow it
>    to be removed from configure.ac

Done so, but I'd appreciate it if you could please double-check that the
way I did it is correct. In particular, I kept the space after directory
name in `configure.ac` but presently I can't remember why this is
important and why this needs to be made so  complicated?

>    In [14]configure.ac:
> > @@ -61,6 +66,16 @@ instrumented_runtime=true
>  force_instrumented_runtime=false
>  instrumented_runtime_libs=""
>  bootstrapping_flexdll=false
> +flexdll_dirs=[[]]
> +lib_dynlink=false
> +lib_str=false
> +lib_systhreads=false
> +lib_unix=false
> +ocamltest_libunix=None
> 
>    There's a bit of a problem with this, and I'm not sure which solution
>    you'll want to do. At the moment, this variable should be None if the
>    Unix library is not built (as here) but then Some true for Unix systems
>    and Some false for Windows systems (the Some false value is missing).
>    This is used to implement the libunix and libwin32unix actions in the
>    testsuite. The simplest fix here is to implement the Some false case,
>    but the reality is that it would be better to eliminate the libunix and
>    libwin32unix actions, sticking with hasunix followed by windows or
>    not-windows as required. What do you think?

Well spotted again! Thanks!

As you may have guessed, I prefer the most conservative road. Because of
the size this PR already has, and because going further is not directly
related to the topic of the PR.

So, I moved the computation of `ocamltest_libunix` close to the one of
`unix_or_win32` since that's the way it's computed on trunk.

>    In [15]configure.ac:
> > +ocamltest_unix_name=""
> +ocamltest_unix_path=""
> +ocamltest_unix_include=""
> 
>    For my taste, these three separate variables go a little bit too far?
>    Isn't it OK for ocamltest/Makefile to know that the Unix library's
>    basename is unix and that if the path is given, we need to include it?
>    In that case, ocamltest_unix_path could be set to '/otherlibs/unix'
>    (note no $(ROOTDIR)) or '' and ocamltest/Makefile could use that single
>    value to instantiate -I $(ROOTDIR)$(ocamltest_unix_path)?

Can such a simplification be left out at the moment? It could go in the
same PR as the previous one, for instance, to be opened once the present
one is merged and, obviously, before `ocamltest/Makefile` also gets
merged in the root Makefile. Would that be okay?

>    In [16]configure.ac:
> >  AS_IF([test x"$enable_unix_lib" != "xno"],
>    [enable_unix_lib=yes
> -   otherlibraries="$otherlibraries unix"])
> +   otherlibraries="$otherlibraries unix"
> +   lib_unix=true
> +   ocamltest_libunix="Some true"
> +   ocamltest_unix_impl="real"
> +   ocamltest_unix_name="unix"
> +   ocamltest_unix_path='$(ROOTDIR)/otherlibs/unix'
> +   ocamltest_unix_include='-I $(ROOTDIR)/otherlibs/unix $(EMPTY)'])
> 
>    If these are kept, the $(EMPTY) could go in Makefile.build_config.in
>    (with no space before it) and the trailing space be left here, which
>    removes more make macros from configure.ac. $(ROOTDIR) could be added
>    to the others using $(ocamltest_unix_path:%=$(ROOTDIR)%) (which is
>    empty if $(ocamltest_unix_path) is empty, but otherwise prefixes it
>    with $(ROOTDIR).

Same here. Can this wait?

>    In [17]configure.ac:
> > @@ -61,6 +66,16 @@ instrumented_runtime=true
>  force_instrumented_runtime=false
>  instrumented_runtime_libs=""
>  bootstrapping_flexdll=false
> +flexdll_dirs=[[]]
> 
>    Is there a subtle reason for using m4sh escaping ([[]]) rather than sh
>    escaping ('[]')?

It's been a long time now since I did that, so I may be wrong but, as
far as I can remember, it simply didn't work. the implementaiton you see
was definitely not natural for me and I had to try several things until
I achieved something that was doing what I wanted. As far as I can
remember, `[]` was interpreted by `m4` as an empty string, whereas, as
far as I understood, what was required was an OCaml empty list...

>    In [18]utils/config_fragment.ml.in:
> > +let with_frame_pointers = @frame_pointers@
> +let profinfo = @profinfo@
> +let profinfo_width = @profinfo_width@
> +
> +let ext_exe = ***@***.***@***@***.***@***@***.***@}
> +let ext_obj = "." ^ ***@***.***@***@***.***@***@***.***@}
> +let ext_asm = "." ^ ***@***.***@***@***.***@***@***.***@}
> +let ext_lib = "." ^ ***@***.***@***@***.***@***@***.***@}
> +let ext_dll = "." ^ ***@***.***@***@***.***@***@***.***@}
> +
> +let host = ***@***.***@***@***.***@***@***.***@}
> +let target = ***@***.***@***@***.***@***@***.***@}
> +
> +let systhread_supported = @systhread_support@
> +
> +let flexdll_dirs = @flexdll_dirs@
> 
>    This is always a list, so it might be clearer to initialise
>    flexdll_dirs to "" in configure.ac and then set it '"+flexdll"' when
>    bootstrapping and here have ***@***.***_dirs@] - or do the square
>    brackets do something strange with the expander??

I am unsure what to do here. Sorry. I think that's one thing we should
discuss directly.

>    In [19]Makefile.build_config.in:
> > @@ -55,15 +55,17 @@ ***@***.***_tool_cmd@
>  # Git submodule)
>  ***@***.***@
>  ***@***.***_flexdll@
> ***@***.***_dir@
> 
>    This isn't needed (if it's kept, it needs the 's' added to
>    @flexdll_dirs@)

I also would like to double check this with yo when we talk to each
other.

>    In [20]utils/config_fragment.ml.in:
> > +  if Sys.win32 || Sys.cygwin && supports_shared_libraries then
> +    let flexlink =
> +      let flexlink =
> +        Option.value ~default:"flexlink" (Sys.getenv_opt "OCAML_FLEXLINK")
> +      in
> +      let f i =
> +        let c = flexlink.[i] in
> +        if c = '/' && Sys.win32 then '\\' else c
> +      in
> +      String.init (String.length flexlink) f
> +    in
> +    let flexdll_chain = ***@***.***@***@***.***_chain@***@***.***@} in
> +    let flexlink_flags = ***@***.***@***@***.***_extra_flags@***@***.***@} in
> +    let flags = " -chain " ^ flexdll_chain ^ " " ^ flexlink_flags in
> +    flexlink ^ flags ^ ***@***.***@***@***.***_ldflags_exp@***@***.***@},
> +    flexlink ^ " -exe" ^ flags ^ ***@***.***@***@***.***_ldflags_exp@***@***.***@},
> 
>    This one should be using OC_LDFLAGS, not OC_DLL_LDFLAGS. I think an
>    equivalent @mkexe_ldflags_exp@ is needed?

I'll see, either alone or we can discuss.

>    In [21]configure.ac:
> > @@ -182,6 +217,8 @@ AC_CONFIG_FILES([Makefile.config])
>  AC_CONFIG_FILES([stdlib/sys.ml])
>  AC_CONFIG_FILES([manual/src/version.tex])
>  AC_CONFIG_FILES([manual/src/html_processing/src/common.ml])
> +AC_CONFIG_FILES([ocamltest/ocamltest_config.ml])
> +AC_CONFIG_FILES([utils/config_fragment.ml])
> 
>    Minor file-naming bike-shedding: we have utils/config.fixed.ml already
>    - utils/config.fragment.ml or even utils/config.generated.ml?

Okay, done.

>    In [23]utils/Makefile:
> > @@ -1,127 +0,0 @@
> -#**************************************************************************
> 
>    The SUBST and SUBST_STRING macros in Makefile.common can also be
>    deleted along with this file!

Good idea! Thanks! I also removed the `SED_SUBST` and `OCAML_SUBST`
auxiliary macros which are not needed any longer either, I think.

>    In [24]Makefile.config.in:
> > @@ -97,7 +97,7 @@ ***@***.***@
>  ***@***.***@
> 
>  ### How to tell the C compiler to output an object file
> ***@***.***@
> ***@***.***@ $(EMPTY)
> 
>    The space before $(EMPTY) shouldn't be here (same in OUTPUTEXE on
>    L192)

Fixed, thanks.

>      __________________________________________________________________
> 
>    In [26]ocamltest/ocamltest_config.ml.in:
> >
> -let systhreads = %%systhreads%%
> +let systhreads = @lib_systhreads@
> 
>    @systhread_support@ could be used, and the extra variable
>    lib_systhreads eliminated.

I'm a bit reluctant here because it breaks the `lib_` idiom we have for
other libraries...

>    In [27]configure.ac:
> > @@ -2138,18 +2221,20 @@ AC_CONFIG_COMMANDS_PRE([
>      AS_IF([test -n "${LDFLAGS}"],
>        [for flag in ${LDFLAGS}; do
>          mkdll_ldflags="${mkdll_ldflags} ${mkexe_ldflags_prefix}${flag}"
> -      done])
> +      done
> +      mkdll_ldflags_exp="$mkdll_ldflags"])
> 
>    I think a mkexe_ldflags_exp is needed here, too (see later comment in
>    utils/config_fragment.ml.in. It would want the oc_ldflags part below on
>    L2231 as well but then not the oc_dll_ldflags bit on L2237

As mentionned earlier, it may be good to discuss this directly.

Thanks again for the precious review!

shindere committed:
Compute FLEXDLL_DIR during configure rather than during build

shindere committed:
Build variables adjustments

Add the $(EMPTY) space-preserving suffix to OUTPUTOBJ, OUTPUTEXE and PACKLD
in Makefile.config.in rather than in the configure script.

This makes it possible to substitute these variables in OCaml files
during the configure stage.

This commit also makes the mkdll_ldflags_exp variable available to
the build system and improves its computation to ensure it does not
contain superfluous spaces.

Finally, this commit instanciates CFLAGS and CPPFLAGS during the configure
stage, so that they can be passed to the OCaml configuraiton module.
It becomes thus impossible to override CFLAGS and CPPFLAGS
at build time.

shindere committed:
Generate config.generated.ml at configure rather than build time

This commit also improves the way mkdll_ldflags_exp is computed so that it
contains no spurious spaces.

shindere commented:
I just rebased this on latest trunk, after the merge of #11420.

Rebasing required to resolve three merge conflicts in configure.

That looks as another argument not to have configure versionned, except for
releases: it would simplify those rebases that modify `configure.ac`
or another of the M4 files.

This should be taken into account in #11343

gasche commented:
Note: One simple way to resolve a conflict on a generated file is to re-generate it.

shindere commented:
Gabriel Scherer (2022/09/07 02:37 -0700):
> Note: One simple way to resolve a conflict on a generated file is to
> re-generate it.

Yes, of course, and it's what I did, three times, but it's both manual
and useless.

dra27 reviewed (approved):
[Precheck seems happy](https://ci.inria.fr/ocaml/job/precheck/761/). With some slight adaptations, I persuaded this PR to produce `utils/config.generated.ml` files for the presently disabled Cygwin/MSVC ports and compared these with `utils.config.ml` from 4.14 (along with mingw).

What this revealed is some breakage from #10413 which we didn't pick up on (it affects mingw-w64 somewhat innocuously, but the faults break MSVC and Cygwin).

However, this PR doesn't make those problems on trunk/5.0 any worse, so I propose as a penance to open a PR to fix those to be merged after this one.

Nice work, @shindere! I'll aim not to shed a tear for the loss of a 50 line `sed` command :)

shindere commented:
David Allsopp (2022/09/09 07:28 -0700):
> @dra27 approved this pull request.
>
> [Precheck seems happy](https://ci.inria.fr/ocaml/job/precheck/761/).
> With some slight adaptations, I persuaded this PR to produce
> `utils/config.generated.ml` files for the presently disabled
> Cygwin/MSVC ports and compared these with `utils.config.ml` from 4.14
> (along with mingw).

I am grateful for all the time and care you invested in reviewing this
ugly PR. Many, many thanks, @dra27. In my opinion, for this PR reviewing
was at least as hard as doingit and @dra27 has been a great, meticuluous
and supportive reviewer.

> What this revealed is some breakage from #10413 which we didn't pick
> up on (it affects mingw-w64 somewhat innocuously, but the faults break
> MSVC and Cygwin).

Oops, so sorry. The present PR and #10413 are really tough bits of
code...

> However, this PR doesn't make those problems on trunk/5.0 any worse,
> so I propose as a penance to open a PR to fix those to be merged after
> this one.

Many thanks again. I'll make sure to review the PR you'll open.

> Nice work, @shindere! I'll aim not to shed a tear for the loss of a 50
> line `sed` command :)

:-)

This also means we don't have any `.plp` files in our codebase, which
feels great.

shindere commented:
David Allsopp (2022/09/09 07:33 -0700):
> Merged #11268 into trunk.

Hurra! :)

Octachron commented:
This PR broke bootstrapping: there are some dependencies on  `config_main` that remain even when `IN_COREBOOT_CYCLE` is true which breaks the `coreboot` target.

dra27 commented:
Oops - it's not to do with the `IN_COREBOOT_CYCLE` stuff, it's simply that `utils/config.generated.ml` is erased during partialclean where it should now only be erased during `distclean`. PR incoming.

shindere commented:
David Allsopp (2022/05/24 02:59 -0700):
> > Is it okay to stop generating `utils/config_boot.mli` and `utils/config_main.mli` as the PR does?
> >
> > These files look not so useful since the corresponding implementations won't actually be compiled as such but rather copied to `config.ml` which will then be compiled and checked against `config.mli`.
>
> These files are there so that if you add, say, `Config.my_amazing_new_feature_flag` to `utils/config.mli` and `utils/config.mlp`, then during the build you get:
>
> ```
> File "/home/dra/ocaml/utils/config_boot.ml", line 1:
> Error: The implementation utils/config_boot.ml
>        does not match the interface utils/config_boot.cmi:
>        The value `my_amazing_new_feature_flag' is required but not provided
>        File "/home/dra/ocaml/utils/config_boot.mli", line 262, characters 0-17:
>          Expected declaration
> ```
>
> Otherwise you only find this out when `make bootstrap` is next run
> (hopefully in CI, of course, but nonetheless further down the line).

I restored the generation of those files. Still, I am wondering whether
the same result couldn't be achieved in another, more straightforward
way. For instance could `config_boot.mli` and `ocnfig_main.mli` be
defined to somehow incllude the signature in `config.mli`? Perhaps we
could introduce, in `config.mli`, an `OCaml_Config` module type that
would contain all the specification of the configuration module and that
could then be included in `config.mli` and then from `config_boot.mli`
and `config_main.mli`, or something similar?

I am also wondering about the following dependency from `compilerlibs/Makefile.compilerlibs`:

```
compilerlibs/ocamlcommon.cma: $(COMMON_CMI) $(ALL_CONFIG_CMO) $(COMMON)
```

I understand its purpose, which you clearly explained above, but it
raises two questions to me.

First, this happens only on the bytecode side. IF we one day manage to
compile the native code directly, we will miss this dependency.

Second, since the dependency is not a real one (e.g.,
`compilerlibs/ocamlcommon.cma` does not really depend on
`config_boot.cmo` when we are not bootstrapping), it feels to me it
would be clearer and cleaner to define a proper target to build compiler
libs and that would build all the appropriate objects, rather than to
use make's dependencies to obtain side effects.

If desired I can include modifications in this PR, or leave that for
later, as is preferred.
